proxy/outprocess/src/module_loaders/outprocess_loader.c

```diff
@@ -20,348 +27,699 @@ DEFINE_ENUM_STRINGS(OUTPROCESS_LOADER_ACTIVATION_TYPE, OUTPROCESS_LOADER_ACTIVAT
+        LogError("Unable to spawn child process(es)!");
+        result = __LINE__;
+    }
+    else
+    {
+        /* Codes_SRS_OUTPROCESS_LOADER_27_049: [** If no errors are encountered, then `OutprocessLoader_SpawnChildProcesses` shall return zero. ] */
+        result = 0;
+    }
+
+    return result;
+}
+
+void OutprocessLoader_JoinChildProcesses(void) {
+    /* Codes_SRS_OUTPROCESS_LOADER_27_064: [ `OutprocessLoader_JoinChildProcesses` shall get the count of child processes, by calling `size_t VECTOR_size(VECTOR_HANDLE handle)`. ] */
+    const size_t child_count = VECTOR_size(uv_processes);
+
+    TICK_COUNTER_HANDLE ticks = NULL;
+    tickcounter_ms_t started_waiting;
+    int uv_thread_result = 0;
+    bool timed_out;
+
+    if (uv_loop_alive(uv_default_loop()))
+    {
+        /* Codes_SRS_OUTPROCESS_LOADER_27_051: [ `OutprocessLoader_JoinChildProcesses` shall create a timer to test for timeout, by calling `TICK_COUNTER_HANDLE tickcounter_create(void)`. ] */
+        if (NULL == (ticks = tickcounter_create()))
+        {
+            /* Codes_SRS_OUTPROCESS_LOADER_27_052: [ If unable to create a timer, `OutprocessLoader_JoinChildProcesses` shall abandon awaiting the grace period. ] */
+            timed_out = true;
+            LogError("failed to create tickcounter");
+        }
+        /* Codes_SRS_OUTPROCESS_LOADER_27_053: [ `OutprocessLoader_JoinChildProcesses` shall mark the begin time, by calling `int tickcounter_get_current_ms(TICK_COUNTER_HANDLE tick_counter, tickcounter_ms_t * current_ms)` using the handle called by the previous call to `tickcounter_create`. ] */
+        else if (tickcounter_get_current_ms(ticks, &started_waiting))
+        {
+            /* Codes_SRS_OUTPROCESS_LOADER_27_054: [** If unable to mark the begin time, `OutprocessLoader_JoinChildProcesses` shall abandon awaiting the grace period. ] */
+            timed_out = true;
+            LogError("failed to sample tickcounter");
+        }
+        else
+        {
+            // Wait for child to clean up
+            tickcounter_ms_t now = started_waiting;
+            /* Codes_SRS_OUTPROCESS_LOADER_27_058: [ `OutprocessLoader_JoinChildProcesses` shall await the grace period in 100ms increments, by calling `void ThreadAPI_Sleep(unsigned int milliseconds)` passing 100 for `milliseconds`. ] */
+            for (timed_out = true; (now - started_waiting) < uv_process_grace_period_ms; ThreadAPI_Sleep(GRACE_AWAIT_DELAY_MS))
+            {
+                /* Codes_SRS_OUTPROCESS_LOADER_27_055: [ While awaiting the grace period, `OutprocessLoader_JoinChildProcesses` shall mark the current time, by calling `int tickcounter_get_current_ms(TICK_COUNTER_HANDLE tick_counter, tickcounter_ms_t * current_ms)` using the handle called by the previous call to `tickcounter_create`. ] */
+                if (tickcounter_get_current_ms(ticks, &now))
+                {
+                    LogError("failed to sample tickcounter");
```

Comment:
I am assuming that a failure here would want the same behavior as timing out...  Because if we hit this condition, we're going to kill all the spawned processes.  Am I correct?